src/persist-client/src/iter.rs

```diff
@@ -132,76 +127,115 @@ pub(crate) enum ConsolidationPart<T, D> {
-                    cursors.push((cursor.clone(), t, D::decode(d)));
-                }
-                cursor.advance(&part);
+        let mut arrays: Vec<&dyn Array> = Vec::with_capacity(parts.len());
+        let mut interleave = |get_array: fn(&BlobTraceUpdates) -> &dyn Array| {
+            for part in parts {
+                arrays.push(get_array(&part.updates));
             }
+            let out =
+                ::arrow::compute::interleave(arrays.as_slice(), indices).expect("valid references");
+            arrays.clear();
+            out
+        };
+        let keys = interleave(|u| u.records().keys()).as_binary().clone();
+        let vals = interleave(|u| u.records().vals()).as_binary().clone();
+        let records = ColumnarRecords::new(keys, vals, timestamps, diffs);
+        let keep_ext = parts.iter().all(|e| e.updates.structured().is_some());
+        let updates = if keep_ext {
+            let key = interleave(|u| &u.structured().unwrap().key);
+            let val = interleave(|u| &u.structured().unwrap().val);
+            BlobTraceUpdates::Both(records, ColumnarRecordsStructuredExt { key, val })
+        } else {
+            BlobTraceUpdates::Row(records)
+        };
 
-            // Sort the cursors by the data that they point to.
-            let kvt = |(cursor, t, _): &(Cursor, T, D)| {
-                cursor.get(&part).map(move |(k, v, _, _)| (k, v, t.clone()))
-            };
-            cursors.sort_by(|a, b| kvt(a).cmp(&kvt(b)));
+        Columnar {
+            updates,
+            _phantom: Default::default(),
+        }
+    }
+}
 
-            ConsolidationPart::Sorted {
-                part,
-                cursors: cursors.into(),
+impl<T: Timestamp + Codec64 + Lattice, D: Codec64> ConsolidationPart<T, D> {
+    pub(crate) fn from_encoded(
+        part: EncodedPart<T>,
+        force_reconsolidation: bool,
+        metrics: &ColumnarMetrics,
+    ) -> Self {
+        let reconsolidate = part.maybe_unconsolidated() || force_reconsolidation;
+        let updates: Columnar<T, D> = Columnar {
+            updates: part.normalize(metrics),
+            _phantom: Default::default(),
+        };
+        let updates = if reconsolidate {
+            let len = updates.len();
+            let mut tuples: Vec<_> = (0..len).map(|i| (updates.get(i).unwrap(), i)).collect();
+            tuples.sort_by_key(|((kv, t, _d), _i)| (*kv, t.clone()));
+
+            let mut indices = Vec::with_capacity(len);
+            let mut timestamps = Int64Builder::with_capacity(len);
+            let mut diffs = Int64Builder::with_capacity(len);
+            for ((_, t, d), i) in tuples {
+                indices.push((0, i));
+                timestamps.append_value(i64::from_le_bytes(T::encode(&t)));
+                diffs.append_value(i64::from_le_bytes(D::encode(&d)));
             }
+            Columnar::interleave(&[&updates], &indices, timestamps.finish(), diffs.finish())
```

Comment:
This feels a bit gross to me - considering an alternative where we just sort the indices and leave the columnar data alone.

Reply:
Still wrapping my head around all this, but leaving the columnar data along and passing around indexes is what I've been (naively) picturing

Reply:
Cool, pushed this! Benchmarking does suggest it's noticeably faster.